zegami_sdk/collection.py

```diff
@@ -367,6 +370,88 @@ def add_feature_pipeline(self, pipeline_name, steps, source=0, generate_snapshot
+                            'params': {
+                                'target_column': 'weight',
+                                'K': 20,
+                            },
+                        },
+                        {
+                            'action': 'cluster',
+                            'params': {
+                                "out_column_start_order": 1010,
+                                'algorithm_args': {
+                                    'algorithm': 'umap',
+                                    'n_components': 2,
+                                    "n_neighbors": 15,
+                                    "min_dist": 0.5,
+                                    "spread": 2,
+                                    "random_state": 123,
+                                }
+                            }
+                        }
+                    ]
+                },
+            ]
+        """
+        all_nodes = self.node_statuses
+        source_names = [s.name for s in self.sources]
+        feature_pipelines = []
+
+        # nodes sorted by source and feature pipeline name
+        feature_pipelines_nodes = {}
+        for source_name in source_names:
+            feature_pipelines_nodes[source_name] = {}
+
+        for node in all_nodes:
+            node_groups = node.get('node_groups')
+            # check if node_groups contain 'feature_pipeline_'
+            if node_groups and len(node_groups) == 3:
+                # get the source name after 'source_'
+                node_source_name = node_groups[0][7:]
+                # get the source name after 'feature_pipeline_'
+                feature_pipeline_name = node_groups[2][17:]
+                if feature_pipeline_name in feature_pipelines_nodes[node_source_name]:
+                    feature_pipelines_nodes[node_source_name][feature_pipeline_name].append(node)
+                else:
+                    feature_pipelines_nodes[node_source_name][feature_pipeline_name] = [node]
+
+        for source_name in source_names:
+            source_pipelines = feature_pipelines_nodes[source_name]
+            for pipeline_name, nodes in source_pipelines.items():
+                for node in nodes:
+                    if 'cluster' in node['source']:
+                        cluster_params = node['source']['cluster']
+                    if 'mRMR' in node['source']:
+                        mRMR_params = node['source']['mRMR']
+                feature_pipelines.append({
+                    'pipeline_name': pipeline_name,
+                    'source_name': source_name,
+                    'steps': [
+                        {
+                            'action': 'mRMR',
+                            'params': mRMR_params,
+                        },
+                        {
+                            'action': 'cluster',
+                            'params': cluster_params,
+                        }
+                    ]
+                })
```

Comment:
This makes a hard assumption that it's always mRMR > cluster. It'd be a bit more future-proof to just map nodes to steps, (assuming we reliably get the nodes in the correct order):
```
steps = [make_step(node) for node in nodes]
```

Reply:
But nodes are not in correct order as we get them from the node statuses, and there is a mapping node that we won't need to use